examples/platforms/efr32mg12/flash.c

```diff
@@ -31,114 +31,66 @@
  *   This file implements the OpenThread platform abstraction for the non-volatile storage.
  */
 
-#include <openthread-core-config.h>
+#include <string.h>
 
-#include <openthread/config.h>
-#include <openthread/platform/alarm-milli.h>
-
-#include "utils/code_utils.h"
-#include "utils/flash.h"
+#include <openthread/instance.h>
 
 #include "em_msc.h"
 
-// clang-format off
-#define FLASH_DATA_END_ADDR     (FLASH_BASE + FLASH_SIZE)
-#define FLASH_DATA_START_ADDR   (FLASH_DATA_END_ADDR - (FLASH_PAGE_SIZE * SETTINGS_CONFIG_PAGE_NUM))
-// clang-format on
+#define FLASH_PAGE_NUM 4
+#define FLASH_DATA_END_ADDR (FLASH_BASE + FLASH_SIZE)
+#define FLASH_DATA_START_ADDR (FLASH_DATA_END_ADDR - (FLASH_PAGE_SIZE * FLASH_PAGE_NUM))
+#define FLASH_SWAP_SIZE (FLASH_PAGE_SIZE * (FLASH_PAGE_NUM / 2))
 
-static inline uint32_t mapAddress(uint32_t aAddress)
+static inline uint32_t mapAddress(uint8_t aSwapIndex, uint32_t aOffset)
 {
-    return aAddress + FLASH_DATA_START_ADDR;
-}
+    uint32_t address;
 
-static otError returnTypeConvert(int32_t aStatus)
-{
-    otError error = OT_ERROR_NONE;
+    address = FLASH_DATA_START_ADDR + aOffset;
+    ;
 
-    switch (aStatus)
+    if (aSwapIndex)
     {
-    case mscReturnOk:
-        error = OT_ERROR_NONE;
-        break;
-
-    case mscReturnInvalidAddr:
-    case mscReturnUnaligned:
-        error = OT_ERROR_INVALID_ARGS;
-        break;
-
-    default:
-        error = OT_ERROR_FAILED;
+        address += FLASH_SWAP_SIZE;
     }
 
-    return error;
+    return address;
 }
 
-otError utilsFlashInit(void)
+void otPlatFlashInit(otInstance *aInstance)
 {
-    MSC_Init();
-    return OT_ERROR_NONE;
-}
+    OT_UNUSED_VARIABLE(aInstance);
 
-uint32_t utilsFlashGetSize(void)
-{
-    return FLASH_DATA_END_ADDR - FLASH_DATA_START_ADDR;
+    MSC_Init();
```

Comment:
This unlocks the flash for writing, leaving it unlocked. We could introduce a `otPlatFlashLock/Unlock` API, to reduce unintended writes. `otPlatFlashInit` would do nothing in this case. What do you think?

Reply:
Also, instead of the proposed APIs, we could just lock/unlock in `otPlatFlashErase/Write`.

Reply:
I agree we should look into this further. However, given that this PR does not change this behavior, I would prefer to address this in a separate PR. Does that work for you?

Reply:
Yes, of course.